core/apple/src/SourcesApple.kt

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.darwin.NSUIntegerVar
+import platform.posix.uint8_tVar
+import kotlin.concurrent.Volatile
+
+/**
+ * Returns an input stream that reads from this source. Closing the stream will also close this source.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Source.asNSInputStream(): NSInputStream = SourceNSInputStream(this)
+
+@OptIn(InternalIoApi::class, UnsafeNumber::class)
+private class SourceNSInputStream(
+    private val source: Source,
+) : NSInputStream(NSData()) {
+
+    private val isClosed: () -> Boolean = when (source) {
+        is RealSource -> source::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    @Volatile
+    private var status = NSStreamStatusNotOpen
+    private var error: NSError? = null
+        set(value) {
+            status = NSStreamStatusError
+            field = value
+        }
+
+    private var pinnedBuffer: Pinned<ByteArray>? = null
+
+    override fun streamStatus() = if (isClosed()) NSStreamStatusClosed else status
+
+    override fun streamError() = error
+
+    override fun open() {
+        if (status == NSStreamStatusNotOpen) {
+            status = NSStreamStatusOpen
+        }
+    }
+
+    override fun close() {
+        status = NSStreamStatusClosed
```

Comment:
At least for `NSInputStream.inputStreamWithFileAtPath` (well, it was the type of stream for which it was easy to check the case), the stream status remains `NSStreamStatusError` after the error happened and `close()` doesn't change it.
I'm not sure if it's explicitly described somewhere in Apple's documentation, but I would suggest mimicking the behavior of standard streams here.

Reply:
That's a great point. I've noticed there are a number of subtle behaviors that aren't explicitly documented by Apple. It seems the terminal statuses for an `NSStream`, where a stream can no longer be used or updated to another status, are either `NSStreamStatusClosed` or `NSStreamStatusError`. Any temporary or parameter-related error doesn't update the `streamError` property or status.

Along these same lines, we should probably close the underlying `Source` when the `NSInputStream` is terminated do to an error. I fixed this for both `SourceNSInputStream` and `SinkNSOutputStream`.